fix: hooks processing fails when config includes token

amimas opened the pull request:
The hooks processor was recently refactored in #635 . One of the change was to inspect and only update a hook when it is needed so that unnecessary api call/update is avoided. This was done by comparing hooks config with data from gitlab about project hooks.

The above fails when a hook config includes a `token` attribute. It's treated as secret and so it is not available in the gitlab data. The comparison fails and gitlabform produces error processing the project.

This PR fixes this issue by adding additional logic for hooks config with token. In this case an update is always done. Otherwise continue with existing logic of updating only if necessary. Updated test cases to validate it.

Fixes #685

amimas commented:
@TigreModerata - you helped with hooks processor recently on couple occasions. Wondering if you would be able to help with a review of this PR. Thanks!

TigreModerata commented:
> @TigreModerata - you helped with hooks processor recently on couple occasions. Wondering if you would be able to help with a review of this PR. Thanks!

Yes, sure, I'm sorry I was slow with my PR (very busy last week, unexpectedly). Good thing you did it anyway, because I wasn't going to add extra logic, rather I was just going to revert back to updating all hooks no matter what.

TigreModerata reviewed (changes requested):
I would suggest reducing the number of nested if's: usually in python I think more than 3 steps is considered too much. 
As for the situation when a token is included in the config: one possibility would be to add an empty-string token to the gl_hook by hand, so that the rest of the processor can continue and compare the config as before. I think it improves readability and makes the nested-if's simplification easy. I don't see any obvious objections to this right now, but maybe you can correct me.

TigreModerata commented on gitlabform/processors/project/hooks_processor.py:
```diff
@@ -15,52 +15,65 @@ def __init__(self, gitlab: GitLab):
-                    debug("Changed hook to '%s'", changed_hook)
-                elif hook_id and not any(diffs):
-                    debug(f"Hook {hook} remains unchanged")
+                    debug(f"Created hook: {created_hook}")
```
```suggestion
                    debug(f"Created hook: {created_hook}")
                    continue
```

TigreModerata reviewed (changes requested):
I tried to include my code snippets here, but it's unreadable if I suggest code one line at a time (I can't see how to do multiline suggestions).

amimas commented:
> > @TigreModerata - you helped with hooks processor recently on couple occasions. Wondering if you would be able to help with a review of this PR. Thanks!
> 
> Yes, sure, I'm sorry I was slow with my PR (very busy last week, unexpectedly). Good thing you did it anyway, because I wasn't going to add extra logic, rather I was just going to revert back to updating all hooks no matter what.

Hey :wave: . no worries at all. Wasn't expecting anything right away. I had a small opportunity and the idea about the extra logic. Thought I'd give it a shot.

Thanks for the feedbacks btw. I'm a python newbi :stuck_out_tongue: If/when I have some time next, I'll look those. I didn't understand clearly about the latter suggestions. Looking at it from mobile. Will check it out later. Thanks again.

TigreModerata commented:
> Thanks for the feedbacks btw. I'm a python newbi 😛 If/when I have some time next, I'll look those. I didn't understand clearly about the latter suggestions. Looking at it from mobile. Will check it out later. Thanks again.

Yeah, you mentioned you were new to python, that's the only reason I even proposed any changes, the logic and everything was fine as it was, obviously :)
 It's just that too many nested loops are frowned upon: they would slow things down considerably if you were dealing with larger sets of data, plus they make things hard to read (almost as bad a fault as being slow, in python ;) ).  My suggestion is just one option for how to rewrite that chunk with just one level of nesting, to give you an idea, but you can probably find better ways.

amimas commented:
Thanks for the feedbacks and suggestions. They all make sense to me. It is good to avoid nested conditions or loops in general; not just for python, I believe. I ended up moving the config diff logic to a separate function to make the hook processing logics a bit more readable.

When you get a chance, please take another look. Thanks again.

TigreModerata commented:
@amimas  last version looks great to me! :)

amimas commented:
Thanks! Pushed one more small update based on the security tests being reported. It was complaining about the setting the token to an empty value, which makes sense. But, I realized I don't even need to do that. Please have a look if you get a chance.

TigreModerata reviewed (commented):
tiny unimportant comment (if I were to nitpick anything "unpythonic"), just to show that one can write it in one line, but I don't think there is any real advantage

amimas commented on gitlabform/processors/project/hooks_processor.py:
```diff
@@ -15,52 +15,95 @@ def __init__(self, gitlab: GitLab):
+            else iter(())
+        )
+
+        if any(diffs):
```
It's not nitpicking. I'm glad you mentioned it. I do tend to forget about python style syntax. Thanks for pointing this out. It's small, simple, and easier to read instead of going through 4 lines.

amimas commented:
So... I was browsing through the code and realized there's already existing logic in the super class to check whether an entity requires update or not. We didn't need to write our own logic. I didn't know about it 🤦 .

Refactored the code and removed the diff checking logic from hooks processor. Used the existing logic instead.

TigreModerata commented:
> So... I was browsing through the code and realized there's already existing logic in the super class to check whether an entity requires update or not. We didn't need to write our own logic. I didn't know about it 🤦 .
> 
> Refactored the code and removed the diff checking logic from hooks processor. Used the existing logic instead.

gulp... I should've noticed! Well, on the bright side... it had to be done (the check added to the superclass), all the better that it has already :D 
Apologising constantly is annoying, and fruitless, but do know I am slapping myself :D
